expression/builtin_time.go

```diff
@@ -1801,71 +1802,81 @@ type extractFunctionClass struct {
 	baseFunctionClass
 }
 
-func (c *extractFunctionClass) getFunction(ctx context.Context, args []Expression) (builtinFunc, error) {
-	if err := c.verifyArgs(args); err != nil {
+func (c *extractFunctionClass) getFunction(ctx context.Context, args []Expression) (sig builtinFunc, err error) {
+	if err = c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinExtractSig{newBaseBuiltinFunc(args, ctx)}
+
+	datetimeUnits := map[string]struct{}{
+		"DAY":             {},
+		"WEEK":            {},
+		"MONTH":           {},
+		"QUARTER":         {},
+		"YEAR":            {},
+		"DAY_MICROSECOND": {},
+		"DAY_SECOND":      {},
+		"DAY_MINUTE":      {},
+		"DAY_HOUR":        {},
+		"YEAR_MONTH":      {},
+	}
+	isDatetimeUnit := true
+	args[0] = WrapWithCastAsString(args[0], ctx)
+	if _, isCon := args[0].(*Constant); isCon {
+		unit, _, err1 := args[0].EvalString(nil, ctx.GetSessionVars().StmtCtx)
```

Comment:
why not check `NULL` ?

Reply:
1. NULL would not happen since the value of the first argument be checked by parser.
2. getFunction cannot handle NULL, we can only handle NULL in evalXXX, so we ignore NULL here.